x-pack/plugin/ent-search/src/test/java/org/elasticsearch/xpack/application/rules/RuleQueryBuilderTests.java

```diff
@@ -0,0 +1,226 @@
+    }
+
+    public void testIllegalArguments() {
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(new MatchAllQueryBuilder(), null, "rulesetId")
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(new MatchAllQueryBuilder(), generateRandomMatchCriteria(), null)
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(new MatchAllQueryBuilder(), generateRandomMatchCriteria(), "")
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(null, generateRandomMatchCriteria(), "rulesetId")
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(null, Collections.emptyMap(), "rulesetId")
+        );
+        expectThrows(
+            IllegalArgumentException.class,
+            () -> new RuleQueryBuilder(
+                new MatchAllQueryBuilder(),
+                generateRandomMatchCriteria("invalid_value"),
+                "rulesetId"
+            )
+        );
+    }
+
+    public void testFromJson() throws IOException {
+        String query = """
+            {
+              "rule_query": {
+                "organic": {
+                  "term": {
+                    "tag": {
+                      "value": "search"
+                    }
+                  }
+                },
+                "match_criteria": {
+                  "query_string": "elastic"
+                },
+                "ruleset_id": "ruleset1"
+              }
+            }""";
+
+        RuleQueryBuilder queryBuilder = (RuleQueryBuilder) parseQuery(query);
+        checkGeneratedJson(query, queryBuilder);
+
+        assertEquals("ruleset1", queryBuilder.rulesetId());
+        assertEquals(query, "elastic", queryBuilder.matchCriteria().get("query_string"));
+        assertThat(queryBuilder.organicQuery(), instanceOf(TermQueryBuilder.class));
+    }
+
+    /**
+    * test that unknown query names in the clauses throw an error
+    */
+    public void testUnknownQueryName() {
+        String query = "{\"rule_query\" : {\"organic\" : { \"unknown_query\" : { } } } }";
+
+        ParsingException ex = expectThrows(ParsingException.class, () -> parseQuery(query));
+        assertEquals("[1:50] [rule_query] failed to parse field [organic]", ex.getMessage());
+    }
+
+    public void testRewrite() throws IOException {
+        RuleQueryBuilder ruleQueryBuilder = new RuleQueryBuilder(
+            new TermQueryBuilder("foo", 1),
+            Map.of("query_string", "bar"),
+            "baz"
+        );
+        QueryBuilder rewritten = ruleQueryBuilder.rewrite(createSearchExecutionContext());
+        assertThat(rewritten, instanceOf(RuleQueryBuilder.class));
+    }
+
+    @Override
+    protected GetResponse executeGet(GetRequest getRequest) {
```

Comment:
@jimczi and @ioanatia - This is where I appear to be running into issues with this test. When I register the async command, `testToQuery` hands infinitely. It doesn't appear to be getting this document, even though I'm setting the getter. This happens both for a regular Get request like we have here, and I see the same behavior when I change this to the original `GetQueryRulesetAction.Request`. Any ideas how I can get this test to pass?